File: 6_tree/前缀树trie/最大异或前缀树/421. 母题数组中两个数的最大异或值.ts
import { useArrayXORTrie } from './XORTrieArray'

/**
 * @param {number[]} nums  0 <= nums[i] <= 2**31 - 1
 * @return {number}
 * 你可以在 O(n) 的时间解决这个问题吗？
 * @summary
 * 每次candidate取贪心 即越大的异或尽量在高位取1
 */
function findMaximumXOR(nums: number[]): number {
  let res = 0
  const trie = useArrayXORTrie(2 ** 31 - 1)

  for (const num of nums) {
    trie.insert(num)
    res = Math.max(res, trie.search(num))
  }

  return res
}

console.log(findMaximumXOR([3, 10, 5, 25, 2, 8]))
console.log(findMaximumXOR([0]))

console.log(25 ^ 8)
console.log(25 ^ 5)

export {}

File: 6_tree/前缀树trie/最大异或前缀树/421. 母题数组中两个数的最大异或值.py
from typing import List
from ATCBinaryTrie import BinaryTrie


class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        n = len(nums)
        max_log = max(nums).bit_length()
        bt = BinaryTrie(add_query_limit=n, max_log=max_log, allow_multiple_elements=False)
        for num in nums:
            bt.add(num)
        res = 0
        for num in nums:  # 查询最大异或值
            bt.xor_all(num)
            res = max(res, bt.maximum())
            bt.xor_all(num)
        return res


assert Solution().findMaximumXOR([3, 10, 5, 25, 2, 8]) == 28
